crates/re_query_cache/src/query.rs

```diff
@@ -142,6 +120,55 @@ macro_rules! impl_query_archetype {
                 Ok(())
             };
 
+
+            let mut latest_at_callback = |query: &LatestAtQuery, latest_at_cache: &mut crate::LatestAtCache| {
+                re_tracing::profile_scope!("latest_at", format!("{query:?}"));
+
+                let crate::LatestAtCache { per_query_time, per_data_time } = latest_at_cache;
+
+                // Fastest path: we have an entry for this exact query, no need to look
+                // any further.
+                if let Some(query_time_bucket) = per_query_time.get(&query.at) {
+                    return iter_results(&query_time_bucket.read());
+                }
+
+                let (data_time, arch_view) = query_archetype::<A>(store, &query, entity_path)?;
+                // TODO(cmc): actual timeless caching support.
+                let data_time = data_time.unwrap_or(TimeInt::MIN);
+
+                // Fast path: we've run the query and realized that we already have the data for the resulting
+                // _data_ time, so let's use that.
+                if let Some(data_time_bucket) = per_data_time.get(&data_time) {
+                    // We now know for a fact that a query at that data time would yield the same
+                    // results: copy the bucket accordingly so that the next cache hit ends up taking the fastest path.
+                    *per_query_time.entry(data_time).or_default() = std::sync::Arc::clone(&data_time_bucket);
```

Comment:
What about this?
```py
rr.set_time(0)
rr.log("points", rr.Points3D([1, 2, 3]))

# Do first query here: LatestAt(+inf)
# Expected: points=[[1,2,3]] colors=[]

rr.set_time(1)
rr.log_components("points", rr.components.Color(0xFF0000))

# Do second query here: LatestAt(+inf)
# Expected: points=[[1,2,3]] colors=[0xFF0000]

rr.set_time(2)
rr.log_components("points", rr.components.Color(0x0000FF))

# Do third query here: LatestAt(+inf)
# Expected: points=[[1,2,3]] colors=[0x0000FF]
```

One way or another this needs to work, which means either the deduplication or the invalidation logic needs to handle this.
Either way, it also probably means that the returned data time should be the max of all components' times?

In any case, need to add a test to prove that this fails miserably.